hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ipc/Server.java

```diff
@@ -4025,4 +4338,326 @@ protected int getMaxIdleTime() {
+              }});
+
+      InetSocketAddress address = new InetSocketAddress(bindAddress, port);
+      io.netty.channel.Channel channel = Server.bind(Binder.NETTY, bootstrap,
+          address, backlogLength, conf, portRangeConfig);
+      registerAcceptChannel(channel);
+      // If may have been an ephemeral port or port range bind, so update
+      // the thread factories to rename any already created threads.
+      port = ((InetSocketAddress)channel.localAddress()).getPort();
+      listenerFactory.updatePort(port);
+      readerFactory.updatePort(port);
+    }
+
+    @Override
+    public InetSocketAddress getAddress() {
+      return Server.this.getListenerAddress();
+    }
+
+    @Override
+    public void listen(InetSocketAddress addr) throws IOException {
+      registerAcceptChannel(Binder.NETTY.bind(bootstrap, addr, backlogLength));
+    }
+
+    @Override
+    public void registerAcceptChannel(io.netty.channel.Channel channel) {
+      acceptChannels.add(channel);
+      addListenerAddress((InetSocketAddress)channel.localAddress());
+    }
+
+    @Override
+    public void closeAcceptChannels() {
+      acceptChannels.close();
+    }
+
+    @Override
+    public void start() {
+      connectionManager.startIdleScan();
+    }
+
+    @Override
+    public void interrupt() {
+      doStop();
+    }
+
+    @Override
+    public void doStop() {
+      try {
+        //TODO : Add Boolean stopped to avoid double stoppage.
+
+        // closing will send events to the bootstrap's event loop groups.
+        closeAcceptChannels();
+        connectionManager.stopIdleScan();
+        connectionManager.closeAll();
+        // shutdown the event loops to reject all further events.
+        ServerBootstrapConfig config = bootstrap.config();
+        //config.group().shutdownGracefully().awaitUninterruptibly();
+        //config.childGroup().shutdownGracefully().awaitUninterruptibly();
+        acceptors.shutdownGracefully().awaitUninterruptibly();
+        readers.shutdownGracefully().awaitUninterruptibly();
+        listenerFactory.close();
+        readerFactory.close();
+      } catch (IOException ioe) {
+        LOG.warn("Unable to shutdown Netty listener and reader threads : ", ioe);
+      }
+      finally {
+        IOUtils.cleanupWithLogger(LOG, listenerFactory, readerFactory);
+      }
+    }
+  }
+
+  @ChannelHandler.Sharable
+  private class NettyResponder extends ChannelOutboundHandlerAdapter
+      implements Responder {
+    @Override
+    public void start() {}
+    @Override
+    public void interrupt() {}
+    // called by handlers.
+    // TODO: Is queuing required similar to the NioResponder implementation ?
+    @Override
+    public void doRespond(RpcCall call) throws IOException {
+      if (LOG.isDebugEnabled()) {
+        LOG.debug(Thread.currentThread().getName() +
+            ": responding to " + call);
+      }
+      NettyConnection connection = call.connection();
```

Comment:
we should also run the code (benchmark) with -Dio.netty.leakDetectionLevel=advanced to detect any potential memory leak.

Reply:
Sure, Thank you for the pointer, I will look into this.